### ISSUE TITLE
RSA Key Gen *sample* test session, vector set and responses result in a failure to validate

### ISSUE THREAD

**Issue by AlexThurston** [usnistgov/ACVP#780]
**environment**
Demo

**testSessionId**
72596

**vsId**
145089

**Algorithm registration**
```json
{
  "algorithm": "RSA",
  "mode": "keyGen",
  "revision": "1.0",
  "fixedPubExp": "010001",
  "infoGeneratedByServer": true,
  "pubExpMode": "fixed",
  "keyFormat": "standard",
  "capabilities": [
    {
      "randPQ": "B.3.3",
      "properties": [
        {
          "modulo": 2048,
          "primeTest": [
            "tblC2",
            "tblC3"
          ]
        },
        {
          "modulo": 3072,
          "primeTest": [
            "tblC2",
            "tblC3"
          ]
        }
      ]
    }
  ]
}
```

**Endpoint in which the error is experienced**
N/A

A vendor is producing the response file for the vector set but when these responses are pushed up, the validation fails with:
`P does not match; Q does not match; N does not match; D does not match`

I've notified the vendor to rule out an error on their part. 

In parallel, I was doing some digging on my own and realized that when a test session isn't marked as a *sample*, there is a key within the test cases named *deferred*.  I haven't been able to find much documentation on what this *deferred* key indicates other than this issue:
https://github.com/usnistgov/ACVP/issues/444

This got me thinking that perhaps when a test session is marked as a *sample* and thus expected responses are generated and available through the API, this somehow affects the validation when responses are pushed up.  The *deferred* key is not within the vector sets in this case.  In the case of RSA Key Gen, data is deferred (and from my understanding intermediate values are calculated and provided by the vendor in the response).

But if the test session is a sample, those intermediates have been calculated without the vendor in order to provide the expected results.

Is it possible that when the vendor pushes up its own response (with its intermediate values) to this sample test session, the comparisons fail and it fails the test cases?

I realize that I'm drawing a number of conclusions, and it might be a situation where the vendor's algorithm isn't correct, but I thought this would be a valuable theory to rule out.

**Comment by smuellerDD** [comment]
Am Mittwoch, 12. Februar 2020, 02:27:21 CET schrieb Alex Thurston:

Hi Alex,

> **environment**
> Demo
> 
> **testSessionId**
> 72596
> 
> **vsId**
> 145089
> 
> **Algorithm registration**
> ```json
> {
>   "algorithm": "RSA",
>   "mode": "keyGen",
>   "revision": "1.0",
>   "fixedPubExp": "010001",
>   "infoGeneratedByServer": true,
>   "pubExpMode": "fixed",
>   "keyFormat": "standard",
>   "capabilities": [
>     {
>       "randPQ": "B.3.3",
>       "properties": [
>         {
>           "modulo": 2048,
>           "primeTest": [
>             "tblC2",
>             "tblC3"
>           ]
>         },
>         {
>           "modulo": 3072,
>           "primeTest": [
>             "tblC2",
>             "tblC3"
>           ]
>         }
>       ]
>     }
>   ]
> }
> ```
> 
> **Endpoint in which the error is experienced**
> N/A
> 
> A vendor is producing the response file for the vector set but when these
> responses are pushed up, the validation fails with: `P does not match; Q
> does not match; N does not match; D does not match`

I also have seen such errors during development, but usually they come from 
the fact that the data is not fully padded. Note, the data must have exactly 
the expected amount of bytes. If one value is smaller, you have to left-pad it 
with zeros.

The data is big-endian format.
> 
> I've notified the vendor to rule out an error on their part.
> 
> In parallel, I was doing some digging on my own and realized that when a
> test session isn't marked as a *sample*, there is a key within the test
> cases named *deferred*.  I haven't been able to find much documentation on
> what this *deferred* key indicates other than this issue:
> https://github.com/usnistgov/ACVP/issues/444

If a vector is marked as sample, an expected result is provided and it cannot 
be used to obtain a cert. But the vector validation should be identical to the 
regular case.

> 
> This got me thinking that perhaps when a test session is marked as a
> *sample* and thus expected responses are generated and available through
> the API, this somehow affects the validation when responses are pushed up. 
> The *deferred* key is not within the vector sets in this case.  In the case
> of RSA Key Gen, data is deferred (and from my understanding intermediate
> values are calculated and provided by the vendor in the response).
> 
> But if the test session is a sample, those intermediates have been
> calculated without the vendor in order to provide the expected results.
> 
> Is it possible that when the vendor pushes up its own response (with its
> intermediate values) to this sample test session, the comparisons fail and
> it fails the test cases?
> 
> I realize that I'm drawing a number of conclusions, and it might be a
> situation where the vendor's algorithm isn't correct, but I thought this
> would be a valuable theory to rule out.



Ciao
Stephan

**Comment by Kritner** [comment]
@AlexThurston it sounds like you're most of the way there.  When `isSample ` is supplied as `true`, the ACVP server takes on the role of both the ACVP Server (obviously) and the IUT.  In the case of your registration, the ACVP server pre-generates the keys prior to delivering the prompt file.  The keys that the server then expects back are the ones that are supplied as the expected results.  If the IUT were instead to calculate their own keys based on the prompt file, it's values wouldn't match the ACVP server's "expected" values.

The `deferred` key is not something I realized was delivered in any files.  @celic do you recall if that's a mistake to include, or if it was purposeful? I didn't think any other algorithm actually returned a `deferred` property. #444

**Comment by AlexThurston** [comment]
@Kritner This is very helpful.  Just to make sure I understand this correctly, the test cases for vector sets (prompt file) that is produced for my particular algorithm registration are actually empty.

They are a series of JSON objects with only a test case ID.  This presumably means that it's up to the IUT to generate and supply the calculated keys and intermediates along the way and provide those in the results/response.

But, because this test session was created as a sample, ACVP produced its own keys and intermediates (as it acted as the IUT).  

Because of this, when responses are submitted they will obviously not match and therefore fail.

If however, I create another test session that *is not* a sample, the ACVP will not have pre-generated values and will use the data provide from the actual IUT in the response to validate algorithm correctness.

Does that sound correct?

**Comment by AlexThurston** [comment]
Incidentally, the test case look like this:
```
{
          "deferred": true,
          "tcId": 76
}
```

So I am seeing the deferred key when I create a test session without it being a sample.

**Comment by Kritner** [comment]
Sorry for the several month late reply @AlexThurston, yes that sounds correct.  I know some changes were made to RSA just recently around sample generation, and I just confirmed that the registration at the open of the issue does (at least now) pass when using the expected results file.

Let me know if there is anything else.
